qualcommax: ipq807x: add support for TCL HH500V

mkrle opened the pull request:
TCL HH500V also known as Vodafone Gigacube B157 is a dual band 802.11ax 5G NR CPE with an FXS port.

Specifications:
* SoC: Qualcomm IPQ8072A
* RAM: 1024 MB
* NAND flash: 256MB NAND (MT29F2G08ABBGAH4)
* NOR flash: 64MB NOR (MX25U25635F) - only in some variants
* WiFi 5G: Qualcomm QCN5054
* WiFi 2.4G: Qualcomm QCN5024
* Ethernet 2.5G: Qualcomm QCA8081
* Ethernet 1G: Qualcomm QCA8075
* Modem: GosunCn GM800 (Qualcomm Snapdragon X55)
* SIM: 1 nano-SIM card slot
* Buttons: Power, Reset, WPS
* LEDs: Power (B/R), WiFi (B), 4G (B/Y/R), 5G (B/Y/R)
* VoIP: 1 FXS RJ11 port (not supported in OpenWrt)
* Power: 12V, 3A

UART serial console:

* 115200,8N1,1.8V
* Three unmarked test points next to QJ2012A:

 +---------+
 | QJ2012A |
 +---------+
 o TX
 o GND
 o RX
 o
 o
 o

Installation via OEM firmware:

* Use the following process to obtain root ssh access
https://forum.openwrt.org/t/openwrt-support-for-vodafone-gigacube-b157/150371/24
* Connect using root with no password on port 42000
* Optionally for serial bootloader access run :
fw_setenv bootdelay=3
* SCP factory.ubi to /tmp
* Run:
sysupgrade -n /tmp/factory.ubi

Installation via serial console and OEM firmware::

* Use console to access OEM firmware shell
* Proceed with SCP & sysupgrade as described above

Other notes:

* This device uses active partition rotation
* Some versions (TCL branded?) have a NOR chip in addition to NAND
* The above is supported by u-boot patching DT partitions
* DT patching does NOT occur on tftpboot/bootm
* Modem is detected as foxconn-sdx55 by kernel (same VID/PID)
* This works OK-ish and should be improved if we can get OEM modem details

Please note that this PR depends on https://github.com/openwrt/firmware_qca-wireless/pull/97

mkrle committed:
generic: 6.12: backport aw9523 fix can_sleep flag

The GPIO expander is connected via I2C, thus the can_sleep flag has to
be set to true. This fixes spurious "scheduling while atomic" bugs
in the kernel ringbuffer.

Signed-off-by: Milan Krstic <milan.krstic@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/19182
Signed-off-by: Robert Marko <robimarko@gmail.com>

mkrle committed:
qualcommax: ipq807x: add support for TCL LINKHUB HH500V

TCL LINKHUB HH500V also known as Vodafone Gigacube B157 is a dual band
802.11ax 5G NR CPE with an FXS port.

Specifications:
* SoC: Qualcomm IPQ8072A
* RAM: 1024 MB
* NAND flash: 256MB NAND (MT29F2G08ABBGAH4)
* NOR flash: 64MB NOR (MX25U25635F) - only in some variants
* WiFi 5G: Qualcomm QCN5054
* WiFi 2.4G: Qualcomm QCN5024
* Ethernet 2.5G: Qualcomm QCA8081
* Ethernet 1G: Qualcomm QCA8075
* Modem: GosunCn GM800 (Qualcomm Snapdragon X55)
* SIM: 1 nano-SIM card slot
* Buttons: Power, Reset, WPS
* LEDs: Power (B/R), WiFi (B), 4G (B/Y/R), 5G (B/Y/R)
* VoIP: 1 FXS RJ11 port (not supported in OpenWrt)
* Power: 12V, 3A

UART serial console:

* 115200,8N1,1.8V
* Three unmarked test points next to QJ2012A:

 +---------+
 | QJ2012A |
 +---------+
 o TX
 o GND
 o RX
 o
 o
 o

Installation via OEM firmware:

* Use the following process to obtain root ssh access
https://forum.openwrt.org/t/150371/24
* Connect using root with no password on port 42000
* Optionally for serial bootloader access run :
fw_setenv bootdelay=3
* SCP factory.bin to /tmp
* Run:
sysupgrade -n /tmp/factory.bin

Installation via serial console and OEM firmware::

* Use console to access OEM firmware shell
* Proceed with SCP & sysupgrade as described above

Other notes:

* This device uses active partition rotation
* Some versions (TCL branded?) have a NOR chip in addition to NAND
* The above is supported by u-boot patching DT partitions
* DT patching does NOT occur on tftpboot/bootm
* Modem is detected as foxconn-sdx55 by kernel (same VID/PID)
* This works OK-ish and should be improved if we can get OEM modem details

Signed-off-by: Milan Krstic <milan.krstic@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/19182
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented on target/linux/generic/pending-6.12/812-usb-serial-option-add-support-for-GosunCn-GM800.patch:
This needs to go upstream, no reason for it to live in OpenWrt

mkrle replied:
> (not the one on the HH500E)
Now that you mentioned you have the HH500E and the GM800 in it, does it get detected as `foxconn-sdx55` on the PCI bus as well, and does the system hang/reboot on removing the `mhi_pci_generic` module? We can take this offline too, I planned to work on these specifics later on.

aiamadeus replied:
Please test this [commit](https://github.com/aiamadeus/openwrt/commit/412a61e10cd811984ae699c56f0af0fd047fe5d8), if it works I will upstream it.

mkrle replied:
I don't think so, I think it must be hardwired. Gpio 46 when taken low forces it into what I thought was download mode (that's how it was marked in OEM firmware scripts):
```
T:  Bus=03 Lev=01 Prnt=01 Port=00 Cnt=01 Dev#=  4 Spd=480  MxCh= 0
D:  Ver= 2.00 Cls=00(>ifc ) Sub=00 Prot=00 MxPS=64 #Cfgs=  1
P:  Vendor=05c6 ProdID=9008 Rev= 0.00
S:  Manufacturer=Qualcomm CDMA Technologies MSM
S:  Product=QUSB_BULK_CID:0410_SN:3C0EDA80
C:* #Ifs= 1 Cfg#= 1 Atr=a0 MxPwr=  2mA
I:* If#= 0 Alt= 0 #EPs= 2 Cls=ff(vend.) Sub=ff Prot=11 Driver=(none)
E:  Ad=81(I) Atr=02(Bulk) MxPS= 512 Ivl=0ms
E:  Ad=01(O) Atr=02(Bulk) MxPS= 512 Ivl=0ms
```

robimarko commented on target/linux/generic/pending-6.12/813-pinctrl-aw9523-fix-add-can_sleep-flag-for-GPIO-chip.patch:
Has this been sent upstream?

mkrle replied:
No. Similar to the modem patch:

1. I've never sent patches to the kernel directly. I have no means of testing this change other than on my specific device. I did test it successfully on my device.
2. I've found this fix/change buried in this commit https://github.com/openwrt/openwrt/commit/00b194871cd82dfbaf678626e49816ab3a51ece8. The entire driver was since upstreamed, but this change was apparently not.  And no one seems to be complaining about it today even though they did initially on the [forums](https://forum.openwrt.org/t/solved-bug-scheduling-while-atomic-procd-1091-0x00000100/115665)

robimarko commented on package/kernel/linux/modules/other.mk:
```diff
@@ -215,6 +215,21 @@ endef
+  DEPENDS:=@GPIO_SUPPORT +kmod-regmap-i2c
+  KCONFIG:=CONFIG_PINCTRL_AW9523
+  FILES:=$(LINUX_DIR)/drivers/pinctrl/pinctrl-aw9523.ko
+  AUTOLOAD:=$(call AutoLoad,40,pinctrl-aw9523,1)
```
Why does it need to be autoloaded manually?

mkrle replied:
I considered this option as well, but when I realised 43/58 qualcommax DTs use gpio-leds and leds-gpio is a module, then I thought this had to be a module for sure. Happy to go either way whatever is the consensus (including submitting another PR to build leds-gpio into the kernel).

robimarko commented on target/linux/qualcommax/ipq807x/base-files/etc/init.d/modem:
Sorry, but why is this needed?

mkrle replied:
The modem boots slightly slower than the router, so it's not probed on the PCI bus when cold booting. It is probed on subsequent soft reboots because the modem doesn't seem get rebooted during these.

I actually planned to implement rebooting the modem during router boot at some point (if I can get the MHI/PCI driver improved to handle unbind better) because a potential modem hang would require a cold power cycle which would be hard if the device is remote. 

Anyhow, after cold boot the modem does not show up in `/sys/bus/pci` until manual rescan which is why I implemented the script to take care of it. Is there a better way?

aiamadeus replied:
You can change the pcie phy driver to a module (CONFIG_PHY_QCOM_QMP_PCIE).

mkrle replied:
> because the modem doesn't seem to get rebooted during these.

I've proven myself wrong, the modem is rebooted during router soft reboot, makes this unnecessary:

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-hh500v.dts:
```diff
@@ -0,0 +1,421 @@
+		pinctrl-0 = <&led_pins>;
+
+		4g_green {
+			label = "green:4g";
```
No, please no labels, color and function only

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-hh500v.dts:
```diff
@@ -0,0 +1,421 @@
+	};
+
+	button_pins: button-pins {
+		power_button {
```
Lets get rid of these QCA style subnodes, no point in them since you are applying literally the same config for all of these pins

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-hh500v.dts:
```diff
@@ -0,0 +1,421 @@
+	switch_lan_bmp = <ESS_PORT4>;
+	switch_wan_bmp = <ESS_PORT6>;
+	switch_mac_mode = <MAC_MODE_PSGMII>;
+	switch_mac_mode2 = <MAC_MODE_SGMII_CHANNEL0>;
```
`MAC_MODE_SGMII_PLUS`

mkrle replied:
Done (assuming you were referring to `switch_mac_mode2`). Seems to be working fine too.

mkrle commented on package/kernel/linux/modules/other.mk:
```diff
@@ -215,6 +215,21 @@ endef
+define KernelPackage/pinctrl-aw9523
+  SUBMENU:=$(OTHER_MENU)
+  TITLE:=Awinic AW9523/AW9523B I2C GPIO expander pinctrl driver
+  HIDDEN:=1
```
I copied it from another GPIO expander driver kmod package assuming there was a reason behind it. I thought that since this is the first and the only use of the kmod package it could make sense. That said, it seems there is another device that uses this module:
https://github.com/openwrt/openwrt/blob/2b0b35310479a993b785bb16b38f40b80e26ad96/target/linux/ramips/dts/mt7621_tenbay_t-mb5eu-v01.dts#L74
 and because of that we have:
https://github.com/openwrt/openwrt/blob/2b0b35310479a993b785bb16b38f40b80e26ad96/target/linux/ramips/mt7621/config-6.6#L230
and
https://github.com/openwrt/openwrt/blob/2b0b35310479a993b785bb16b38f40b80e26ad96/target/linux/ramips/mt7621/config-6.12#L232
I thought it's best not to touch a such a popular target without being able to test it so I haven't made any changes there. But I did think including the driver in all device images just for one was a bit too much which is why I created the kmod package instead of including it in the kernel (config).

I have removed `HIDDEN` now, please let me know your thoughts on that and the above.

mkrle commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-hh500v.dts:
```diff
@@ -0,0 +1,421 @@
+	};
+
+	gpio-poweroff {
+		compatible = "gpio-poweroff";
```
It's really poweroff, and without it `poweroff` just reboots. It's that way in the OEM firmware and I tested it works properly this way in OpenWrt. 

That said, the modem MHI/PCI driver is not great (yet) so it may reboot the system when unbinding, but that doesn't change the fact that this GPIO is needed for a proper power off. Not that poweroff is a killer feature but still.

aiamadeus commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-hh500v.dts:
```diff
@@ -0,0 +1,405 @@
+#include <dt-bindings/leds/common.h>
+
+/ {
+	model = "TCL HH500V";
```
Should it be "TCL LinkHub HH500V"?

aiamadeus replied:
The model name of HH500E on Amazon is "TCL LinkHub HH500E".

mkrle replied:
> The model name of HH500E on Amazon is "TCL LinkHub HH500E".

Can you please share the link? For example [here](https://amzn.eu/d/dvhBH7D) the title is "TCL LinkHub HH500E 5G CPE" and it under technical details it says "Item model number: ‎HH500E"

aiamadeus replied:
Looks like all capital letters: https://www.tcl.com/global/en/news/tcl-linkhub-hh500v-was-awarded-with-certificate-of-product-carbon-footprint

aiamadeus commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -392,6 +392,22 @@ define Device/spectrum_sax1v1k
+	$(call Device/FitImage)
+	$(call Device/UbiFit)
+	DEVICE_VENDOR := TCL
+	DEVICE_MODEL := HH500V
```
The device tree looks no different from my HH500E.
It seems that you can add HH500E as ALT0_MODEL.
But some HH500V boards may have secure boot enabled?

mkrle replied:
Mine doesn't, and the Vodafone variant from the forum member doesn't either. Would it make any difference in terms of this specific proposed changed if some had it enabled?

Would you want to do any tests on your device or you're confident it's good to go?

aiamadeus replied:
Not yet, I just sent the gm800 patch upstream.

mkrle replied:
Ok, I can also add it later on.

mkrle commented:
Hi @robimarko, just wanted to let you know this one is ready for another review round, at your convenience. Thanks!

robimarko commented on target/linux/qualcommax/ipq807x/base-files/etc/board.d/02_network:
```diff
@@ -101,6 +102,10 @@ ipq807x_setup_macs()
 		;;
+	tcl,hh500v)
+		wan_mac=$(mtd_get_mac_binary "0:ART" 0x5e290)
+		lan_mac=$(mtd_get_mac_binary "0:ART" 0x5e296)
```
You should be able to use NVMEM instead

mkrle replied:
The tricky part is that I am relying on u-boot to populate the partitions in the DTS, and ART partition can be on NOR or NAND depending on the HW. I am not sure if DT would compile if I referenced a partition node that does not (yet) exist, but even if it did I don't know how I would reference it as u-boot does not set partition aliases, and absolute path can be either SPI NOR or NAND.

Any tips on how it can be done?

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-hh500v.dts:
```diff
@@ -0,0 +1,402 @@
+		serial0 = &blsp1_uart5;
+		serial1 = &blsp1_uart3;
+		ethernet3 = &dp4;
+		ethernet5 = &dp6_syn;
```
Considering that you are setting MAC-s manually, you can drop these ethernet aliases

mkrle replied:
Done.

mkrle commented:
I've made a minor change to the factory image so that only one is built. This one can be used for both OEM sysupgrade as well as uboot flashing if needed. Rebased on top of master.

mkrle commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -392,6 +392,20 @@ define Device/spectrum_sax1v1k
+	PAGESIZE := 2048
+	DEVICE_DTS_CONFIG := config@hk09
+	SOC := ipq8072
+	IMAGE/factory.ubi := append-ubi | qsdk-ipq-factory-nand
```
Yes, and the DTB version can be used also from u-boot with `imextract` right? Anyhow, it's not that important, I have reverted this.

mkrle committed:
qualcommax: ipq807x: add functions for bootconfig partition

The BOOTCONFIG partition is used by Qualcomm's boot chain to store
metadata about the device's startup configuration. It contains info such
as versioning, configuration flags, primary boot partition, and more.

Newer devices with dual boot partitions not only store the active boot
partition in a U-boot variable but also in partition info in the
BOOTCONFIG partition. As such, add library functions to set and toggle
the active boot partition.

Signed-off-by: Milan Krstic <milan.krstic@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/19182
Signed-off-by: Robert Marko <robimarko@gmail.com>

mkrle commented:
Probably, though i wouldn't recommend it. The issue was introduced in 6.12.58, so reverting to 6.12.57 would fix it. We're working on a fix for this in OpenWrt and upstream (https://github.com/openwrt/openwrt/pull/22985).
